metagpt/utils/repair_llm_raw_output.py

```diff
@@ -0,0 +1,300 @@
+            # problem, miss char `,` at the end.
+            line = f"{line},"
+
+        arr[line_no] = line
+        output = "\n".join(arr)
+        logger.info(f"repair_invalid_json, raw error: {error}")
+
+    return output
+
+
+def run_after_exp_and_passon_next_retry(logger: "loguru.Logger") -> Callable[["RetryCallState"], None]:
+    def run_and_passon(retry_state: RetryCallState) -> None:
+        """
+        RetryCallState example
+            {
+                "start_time":143.098322024,
+                "retry_object":"<Retrying object at 0x7fabcaca25e0 (stop=<tenacity.stop.stop_after_attempt ... >)>",
+                "fn":"<function retry_parse_json_text_v2 at 0x7fabcac80ee0>",
+                "args":"(\"tag:[/CONTENT]\",)",  # function input args
+                "kwargs":{},                     # function input kwargs
+                "attempt_number":1,              # retry number
+                "outcome":"<Future at xxx>",  # type(outcome.result()) = "str", type(outcome.exception()) = "class"
+                "outcome_timestamp":143.098416904,
+                "idle_for":0,
+                "next_action":"None"
+            }
+        """
+        if retry_state.outcome.failed:
+            if len(retry_state.args) > 0:
+                # # can't used as args=retry_state.args
+                func_param_output = retry_state.args[0]
+            elif len(retry_state.kwargs) > 0:
+                func_param_output = retry_state.kwargs.get("output", "")
+            # import pdb; pdb.set_trace()
+            exp_str = str(retry_state.outcome.exception())
+            logger.warning(f"parse json from content inside [CONTENT][/CONTENT] failed at retry "
+                           f"{retry_state.attempt_number}, try to fix it, exp: {exp_str}")
+
+            repaired_output = repair_invalid_json(func_param_output, exp_str)
+            retry_state.kwargs["output"] = repaired_output
+
+    return run_and_passon
+
+
+@retry(
+    stop=stop_after_attempt(3 if CONFIG.repair_llm_output else 0),
+    wait=wait_fixed(1),
+    after=run_after_exp_and_passon_next_retry(logger),
+)
+def retry_parse_json_text(output: str) -> Union[list, dict]:
+    """
+    repair the json-text situation like there are extra chars like [']', '}']
+
+    Warning
+        if CONFIG.repair_llm_output is False, retry _aask_v1 {x=3} times, and the retry_parse_json_text's retry not work
+        if CONFIG.repair_llm_output is True, the _aask_v1 and the retry_parse_json_text will loop for {x=3*3} times.
+            it's a two-layer retry cycle
+    """
+    logger.debug(f"output to json decode:\n{output}")
+
+    # if CONFIG.repair_llm_output is True, it will try to fix output until the retry break
+    parsed_data = CustomDecoder(strict=False).decode(output)
+
+    return parsed_data
+
+
+def extract_content_from_output(content: str, right_key: str = "[/CONTENT]"):
+    """ extract xxx from [CONTENT](xxx)[/CONTENT] using regex pattern """
```

Comment:
@shenchucheng you can have a check about the extraction pattern.
The common cases are list in the `test_repair_llm_raw_output.py`, there is one case can't pass with original extract pattern, so I change it. But I think it's not good enough.

Target: find a regex pattern can handle most cases of generated output.